perf: optimize spawn cleanup removing redundant iteration

kaleohanopahala opened the pull request:
This improves efficiency by avoiding an intermediate list for removal, saving time and memory. The map is traversed only once, with iterators updated after each removal, making the process more efficient. Additionally, it eliminates redundant lookups by directly modifying the map during iteration.

kaleohanopahala committed:
Update spawn_monster.cpp

kaleohanopahala committed:
Fix mistake

ops.

luanluciano93 commented:
you have to set **lastSpawn**

mehah commented:
In your original PR, you used erase_if, why did you give up using it, it does the same thing but is visually cleaner.
https://github.com/opentibiabr/canary/pull/2907#issuecomment-2373061460

![image](https://github.com/user-attachments/assets/8255ab92-d796-44ae-a46f-57fc05a192bd)

I believe the overhead of calling a lambda is irrelevant.

kaleohanopahala committed:
Update spawn_monster.cpp

Fixed Lastspawn
Add nullcheck.

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala commented:
> you have to set **lastSpawn**

Sorry,  I didn't fully understand how it worked... 🥲
Now corrected the missing lastSpawn and added a nullity check.

Working well in my test.

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

murilo09 commented:
Can you please check the issue in Sonar?

kaleohanopahala committed:
fix sonar and c++20 code

kaleohanopahala committed:
Merge branch 'main' into kaleohanopahala-patch-1

kaleohanopahala commented:
> Can you please check the issue in Sonar?

Ya, dude.

dudantas commented:
The change to use std::erase_if directly poses risks of undefined behavior and potential crashes because it modifies the container while iterating over it. This can lead to invalid references, iterator corruption, or segmentation faults if elements are removed concurrently. To ensure stability and avoid these risks, we need a safe mechanism that guarantees no modifications are made while traversing the container, such as first collecting keys to remove before performing any deletions.

[crash.txt](https://github.com/user-attachments/files/17400420/crash.txt)

kaleohanopahala commented:
> The change to use std::erase_if directly poses risks of undefined behavior and potential crashes because it modifies the container while iterating over it. This can lead to invalid references, iterator corruption, or segmentation faults if elements are removed concurrently. To ensure stability and avoid these risks, we need a safe mechanism that guarantees no modifications are made while traversing the container, such as first collecting keys to remove before performing any deletions.
> 
> [crash.txt](https://github.com/user-attachments/files/17400420/crash.txt)

Hmm...

Edit: Can you try to crash now?
Now looks good.

```
void SpawnMonster::cleanup() {
	std::vector<int> keysToRemove;

	for (const auto &entry : spawnedMonsterMap) {
		const auto &monster = entry.second;
		if (!monster || monster->isRemoved()) {
			auto spawnIt = spawnMonsterMap.find(entry.first);
			if (spawnIt != spawnMonsterMap.end()) {
				spawnIt->second.lastSpawn = OTSYS_TIME();
			}
			keysToRemove.push_back(entry.first);
		}
	}

	for (const auto &key : keysToRemove) {
		spawnedMonsterMap.erase(key);
	}
}
```

Edit: I noticed that you don't use 'keysToRemove'; however, in this approach, I didn't encounter performance issues. I made another one that followed your model, but it generated instabilities with a high flow. If you could provide feedback, I'll take a look later :D
